### PR TITLE
Add kpm wrap assembly

### PR DESCRIPTION
**Pull request by ChengTian**
parent https://github.com/aspnet/DNX/issues/1385 https://github.com/aspnet/DNX/issues/1316

- Automatically wrap dependency assemblies in same dir
- Automatically wrap pdb in same dir

Functional tests will be added soon. This PR allows people help to review usage ASAP.

@davidfowl @lodejard @muratg

### CODE REVIEW

**Comment by msotclas** [comment]
Hi __@ChengTian__, I'm your friendly neighborhood Microsoft Open Technologies, Inc. Pull Request Bot (You can call me MSOTBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Wei Wang (ASP.NET)). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla2.msopentech.com.
    </p>

TTYL, MSOTBOT;

**Comment by ChengTian** [comment]
Sample usage:
wrapassembly

Then you get:
(1) `<RootDir>/wrap/ClassLibrary1/project.json`:
```json
{
  "version": "1.0.0-*",
  "frameworks": {
    "dnx451": {
      "bin": {
        "assembly": "../../../libs/ClassLibrary1.dll",
        "pdb": "../../../libs/ClassLibrary1.PDB"
      },
      "dependencies": {
        "ClassLibrary2": "1.0.0-*",
        "CLASSLIBRARY3": "1.0.0-*"
      }
    }
  }
}
```

(2) `<RootDir>/wrap/ClassLibrary2/project.json`:
```json
{
  "version": "1.0.0-*",
  "frameworks": {
    "dnx451": {
      "bin": {
        "assembly": "../../../libs/ClassLibrary2.dll",
        "pdb": "../../../libs/CLASSLibrary2.pdb"
      }
    }
  }
}
```

(3) `<RootDir>/wrap/CLASSLIBRARY3/project.json`:
```json
{
  "version": "1.0.0-*",
  "frameworks": {
    "dnx451": {
      "bin": {
        "assembly": "../../../libs/CLASSLIBRARY3.DLL",
        "pdb": "../../../libs/CLASSLIBRARY3.PDB"
      }
    }
  }
}
```

File: src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs [hunk 1]
```diff
@@ -416,25 +452,68 @@ private static bool IsFrameworkAssembly(XElement itemElement)
...
-            AddFrameworkBinPaths(projectJson, relativeAssemblyPath, targetFramework, addPdbPath: false);
+            var pdbPath = GetSideBySidePdbPath(assemblyPath);
+            var relativePdbPath = PathUtility.GetRelativePath(targetProjectJson, pdbPath, PathSeparator);
+
+            var projectJson = LoadOrCreateProjectJson(targetProjectJson);
+            AddFrameworkBinPaths(projectJson, targetFramework, relativeAssemblyPath, relativePdbPath);
+
+            // Wrap dependency assemblies that are in the same dir
+            var assemblyInfo = new AssemblyInformation(assemblyPath, processorArchitecture: null);
+            var dependencyAssemblies = assemblyInfo.GetDependencies().Select(x => x + ".dll");
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: Hashset?

**Comment by ChengTian** [comment]
Made code review changes. Also fixed one more bug, the current behavior is:
Assume `AssemblyA.dll` has a dependency `AssemblyB.dll` in its dll metadata.
When `kpm wrap AssemblyA.dll` and we find a sibling assembly with name `ASSEMBLYB.DLL`, we treat it as a dependency and also generate a wrapper project for it. The generated wrapper project is `AssemblyB` instead of `ASSEMBLYB`. That means, we detect dependency assembly with case-insensitive string comparison but the wrapper project's name is determined by metadata in dependee.

**Comment by ChengTian** [comment]
The final behavior above is consistent with dll reference operation in VS2013. In VS2013, assume `AssemblyA.dll` has a dependency `AssemblyB.dll` in its metadata. When adding assembly reference to `AssemblyA.dll`, if VS2013 finds `ASSEMBLYB.DLL` in the same dir, it will also copy `ASSEMBLYB.DLL` to local bin dir and rename it to `AssemblyB.dll`. So the dependency detection is case-insensitive and final dependency name is determined by metadata of AssemblyA.dll.

**Comment by ChengTian** [comment]
Let's check this in if reviewers give sign-off so that Tooling team can work on this feature ASAP. Adding functional tests for this can be tracked at https://github.com/aspnet/dnx/issues/1415

**Comment by davidfowl** [comment]
What about `kpm bundle` and `kpm bundle --no-source` operations?

**Comment by ChengTian** [comment]
@davidfowl , I guess you are talking about this: https://github.com/aspnet/dnx/pull/1136 (kpm bundle automatically "no-source" wrapper project)

**Comment by davidfowl** [comment]
@ChengTian Yes that has to go in as well.

**Comment by davidfowl** [comment]
@ChengTian did you test it on mono?

**Comment by ChengTian** [comment]
No, because MSBuild is not available on *nix, we made `kpm wrap` unavailable on Mono: https://github.com/aspnet/dnx/blob/dev/src/Microsoft.Framework.PackageManager/Program.cs#L505

But, yeah, we can divide `kpm wrap` into two pieces, `kpm wrap <csproj>` is only available on Windows, while `kpm wrap <assembly>` is available on all platforms. I'll add that part soon.

**Comment by ChengTian** [comment]
@davidfowl , made `kpm wrap <assembly>` available on Mono. Also did verification on Mono and it works well.

**Comment by ChengTian** [comment]
Let's get this in if reviewers find Ok. I think Tooling team wants this feature ASAP. @davidfowl @muratg @lodejard

**Comment by davidfowl** [comment]
:shipit:

**Pull request merged**